src/ParallelAlgorithms/Events/ObserveVolumeIntegrals.hpp

```diff
@@ -0,0 +1,202 @@
+               ? ""
+               : "_" + tensor.component_name(
+                           tensor.get_tensor_index(component_index));
+  }
+
+ public:
+  /// \cond
+  explicit ObserveVolumeIntegrals(CkMigrateMessage* /*unused*/) noexcept {}
+  using PUP::able::register_constructor;
+  WRAPPED_PUPable_decl_template(ObserveVolumeIntegrals);  // NOLINT
+  /// \endcond
+
+  using options = tmpl::list<>;
+  static constexpr OptionString help =
+      "Observe the volume integrals of the tensors over the domain.\n"
+      "\n"
+      "Writes reduction quantities:\n"
+      " * ObservationValueTag\n"
+      " * Volume = volume of the domain\n"
+      " * VolInt(*) = volume integral of the tensor\n"
+      "\n"
+      "Warning: Currently, only one reduction observation event can be\n"
+      "triggered at a given observation value.  Causing multiple events to\n"
+      "run at once will produce unpredictable results.";
+
+  ObserveVolumeIntegrals() = default;
+
+  using observed_reduction_data_tags =
+      observers::make_reduction_data_tags<tmpl::list<ReductionData>>;
+
+  using argument_tags =
+      tmpl::list<ObservationValueTag, domain::Tags::Mesh<VolumeDim>,
+                 domain::Tags::ElementMap<VolumeDim, Frame::Inertial>,
+                 domain::Tags::Coordinates<VolumeDim, Frame::Logical>,
+                 Tensors...>;
+
+  template <typename Metavariables, typename ArrayIndex,
+            typename ParallelComponent>
+  void operator()(
+      const db::const_item_type<ObservationValueTag>& observation_value,
+      const Mesh<VolumeDim>& mesh,
+      const ElementMap<VolumeDim, Frame::Inertial>& element_map,
+      const tnsr::I<DataVector, VolumeDim, Frame::Logical>& logical_coordinates,
+      const db::const_item_type<Tensors>&... tensors,
+      Parallel::ConstGlobalCache<Metavariables>& cache,
+      const ArrayIndex& /*array_index*/,
+      const ParallelComponent* const /*meta*/) const noexcept {
+    // Determinant of Jacobian is needed because integral is performed in
+    // logical coords. Currently not initialized in the Metavariables.
+    const DataVector det_jacobian =
+        get(determinant(element_map.jacobian(logical_coordinates)));
+    const double local_volume = definite_integral(det_jacobian, mesh);
+
+    std::vector<double> local_vol_ints{};
+    std::vector<std::string> reduction_names = {
+        db::tag_name<ObservationValueTag>(), "Volume"};
+    const auto record_integrals =
+        [&local_vol_ints, &reduction_names, &det_jacobian, &mesh ](
+            const auto tensor_tag_v, const auto& tensor) noexcept {
+      using tensor_tag = tmpl::type_from<decltype(tensor_tag_v)>;
+      for (size_t i = 0; i < tensor.size(); ++i) {
+        reduction_names.push_back("VolInt(" + db::tag_name<tensor_tag>() +
+                                  component_suffix(tensor, i) + ")");
+        local_vol_ints.push_back(
+            definite_integral(det_jacobian * tensor[i], mesh));
```

Comment:
[not a change request] Just for the record, in the future we will want to think about and figure out how to support different types of norms, ideally controlled by the input.